cockatrice/src/handle_public_servers.cpp

```diff
@@ -0,0 +1,92 @@
+
+HandlePublicServers::HandlePublicServers(QObject *parent) : QObject(parent), nam(nullptr), reply(nullptr)
+{
+}
+
+void HandlePublicServers::downloadPublicServers()
+{
+    QUrl url(QString(PUBLIC_SERVERS_JSON));
+    nam = new QNetworkAccessManager(this);
+    reply = nam->get(QNetworkRequest(url));
+    connect(reply, SIGNAL(finished()), this, SLOT(actFinishParsingDownloadedData()));
+}
+
+void HandlePublicServers::actFinishParsingDownloadedData()
+{
+    reply = dynamic_cast<QNetworkReply *>(sender());
+    QNetworkReply::NetworkError errorCode = reply->error();
+
+    // After finished() is called, this object will be deleted
+    nam->deleteLater();
+
+    if (errorCode == QNetworkReply::NoError) {
+        // Get current saved hosts
+        UserConnection_Information uci;
+        savedHostList = uci.getServerInfo();
+
+        // List of public servers
+        // Will be populated below
+        QStringList publicServersAvailable;
+        QStringList publicServersToRemove;
+
+        // Downloaded data from GitHub
+        QString jsonData = QString(reply->readAll());
+        auto jsonMap = QtJson::Json::parse(jsonData).toMap();
+
+        // Servers available
+        auto publicServersJSONList = jsonMap["servers"].toList();
+        for (const auto &server : publicServersJSONList) {
+            // Data inside one server at a time
+            // server: [{ ... }, ..., { ... }]
+            const auto serverMap = server.toMap();
+
+            QString serverName = serverMap["name"].toString();
+
+            if (serverMap["isInactive"].toBool()) {
+                publicServersToRemove.append(serverName);
+                continue;
+            }
+
+            QString serverAddress = serverMap["host"].toString();
+            QString serverPort = serverMap["port"].toString();
+
+            publicServersAvailable.append(serverName);
+
+            if (!savedHostList.contains(serverName)) {
+                settingsCache->servers().addNewServer(serverName, serverAddress, serverPort, "", "", false);
+            } else {
+                settingsCache->servers().updateExistingServerWithoutLoss(serverName, serverAddress, serverPort);
+            }
+        }
+
+        // If a server was removed from the public list,
+        // we will delete it from the local system.
+        // Will not delete "unofficial" servers
+        for (auto server : savedHostList) {
+            QString serverName = server.getSaveName();
+            bool isCustom = server.isCustomServer();
+
+            if (!isCustom && publicServersToRemove.indexOf(serverName) != -1) {
+                settingsCache->servers().removeServer(serverName);
+            }
+        }
+
+        emit sigPublicServersDownloadedSuccessfully();
+    } else {
+        qDebug() << "[CONNECTION DIALOG]"
+                 << "Error Downloading Public Servers" << errorCode;
+        emit sigPublicServersDownloadedUnsuccessfully();
```

Comment:
Include the error code?

Reply:
Sure... doesn't really matter as qdebug says it, but i can include it